File: data/derived_cot/rq3_prompting/results/translation/r1/cpp_to_py/code_output/SQLQueryBuilder.py
from typing import List, Tuple, Optional

class SQLQueryBuilder:
    @staticmethod
    def select(
        table: str,
        columns: Optional[List[str]] = None,
        where: Optional[List[Tuple[str, str]]] = None
    ) -> str:
        if columns is None:
            columns = ["*"]
        if where is None:
            where = []

        if len(columns) == 1 and columns[0] == "*":
            query = "SELECT *"
        else:
            query = "SELECT " + ", ".join(columns)

        query += f" FROM {table}"

        if where:
            conditions = " AND ".join(f"{k}='{v}'" for k, v in where)
            query += f" WHERE {conditions}"

        return query

    @staticmethod
    def insert(
        table: str,
        data: List[Tuple[str, str]]
    ) -> str:
        columns = ", ".join(k for k, v in data)
        values = ", ".join(f"'{v}'" for k, v in data)
        return f"INSERT INTO {table} ({columns}) VALUES ({values})"

    @staticmethod
    def delete_(
        table: str,
        where: Optional[List[Tuple[str, str]]] = None
    ) -> str:
        if where is None:
            where = []

        query = f"DELETE FROM {table}"
        if where:
            conditions = " AND ".join(f"{k}='{v}'" for k, v in where)
            query += f" WHERE {conditions}"

        return query

    @staticmethod
    def update(
        table: str,
        data: List[Tuple[str, str]],
        where: Optional[List[Tuple[str, str]]] = None
    ) -> str:
        if where is None:
            where = []

        set_clause = ", ".join(f"{k}='{v}'" for k, v in data)
        query = f"UPDATE {table} SET {set_clause}"
        if where:
            conditions = " AND ".join(f"{k}='{v}'" for k, v in where)
            query += f" WHERE {conditions}"

        return query

File: data/derived_cot/rq1_traces/translation/qwen/py_to_cpp/code_output/SQLQueryBuilder.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <map>
#include <optional>

class SQLQueryBuilder {
public:
    static std::string select(const std::string& table, const std::vector<std::string>& columns = {"*"}, std::optional<std::map<std::string, std::string>> where = std::nullopt) {
        std::stringstream query;
        query << "SELECT ";
        if (columns[0] == "*") {
            query << "*";
        } else {
            for (size_t i = 0; i < columns.size(); i++) {
                if (i > 0) query << ", ";
                query << columns[i];
            }
        }
        query << " FROM " << table;
        if (where) {
            query << " WHERE ";
            for (const auto& [key, value] : *where) {
                if (&key != &(*where.begin()->first)) {
                    query << " AND ";
                }
                query << key << "='" << value << "'";
            }
        }
        return query.str();
    }

    static std::string insert(const std::string& table, const std::map<std::string, std::string>& data) {
        std::stringstream query;
        query << "INSERT INTO " << table << " (";
        for (const auto& [key, value] : data) {
            if (&key != &data.begin()->first) {
                query << ", ";
            }
            query << key;
        }
        query << ") VALUES (";
        for (const auto& [key, value] : data) {
            if (&key != &data.begin()->first) {
                query << ", ";
            }
            query << "'" << value << "'";
        }
        query << ")";
        return query.str();
    }

    static std::string delete_(const std::string& table, std::optional<std::map<std::string, std::string>> where = std::nullopt) {
        std::stringstream query;
        query << "DELETE FROM " << table;
        if (where) {
            query << " WHERE ";
            for (const auto& [key, value] : *where) {
                if (&key != &(*where.begin()->first)) {
                    query << " AND ";
                }
                query << key << "='" << value << "'";
            }
        }
        return query.str();
    }

    static std::string update(const std::string& table, const std::map<std::string, std::string>& data, std::optional<std::map<std::string, std::string>> where = std::nullopt) {
        std::stringstream query;
        query << "UPDATE " << table << " SET ";
        for (const auto& [key, value] : data) {
            if (&key != &data.begin()->first) {
                query << ", ";
            }
            query << key << "='" << value << "'";
        }
        if (where) {
            query << " WHERE ";
            for (const auto& [key, value] : *where) {
                if (&key != &(*where.begin()->first)) {
                    query << " AND ";
                }
                query << key << "='" << value << "'";
            }
        }
        return query.str();
    }
};